### ISSUE TITLE
tpmtest #34 fails with error 0x8006, while succeeds if runs #34 alone

### ISSUE THREAD

**Issue by yunyings** [01org/TPM2.0-TSS#232]
Hardware Platform: Intel x86_64
TPM Type: Intel PTT (INTC)
OS: Red Hat Enterprise Linux Client 7.2(Maipo) 
Kernel: mainline kernel 4.7.0-rc4+
Application:
	TPM2.0-TSS	

Error description:
Runs ./TPM2.0-TSS/test/tpmtest/tpmtest, to test TPM on Intel x86_64. Test #34(ASYMMETRIC ENCRYPT/DECRYPT & HASH TESTS) always fails, with error below:
FAILED!  System API Error: 0x80006

However, if runs #34 test alone, test always pass.

Problem doesn't seem to be at kernel side, since 4.5-rc4 is the latest mainline Linux kernel.
Is there any unidentified problem with TPM2.0-TSS tpmptest itself?

I can provide more detailed information if needed.

**Comment by yunyings** [comment]
Running test #34 alone - **PASS**:
Please select an action:34

ASYMMETRIC ENCRYPT/DECRYPT TESTS:
Q - QUIT THIS TEST GROUP
D - PRINT DESCRIPTION ON ALL CASES IN THIS GROUP
0 - RUN ALL TEST CASES
Please select an action:0

RUN ALL TEST CASES:

ASYMMETRIC ENCRYPT/DECRYPT & HASH TESTS:
	passing case:  	PASSED!

New key successfully created in owner hierarchy (RSA 2048).  Handle: 0x80000004
Name of created primary key: 00 0b 4f 63 f6 78 f9 44 70 1f 69 c0 d3 10 eb 07 
a1 d7 7e 3c cf b5 20 e6 4e 43 20 a4 e8 75 e2 68 
30 f4 
	passing case:  	PASSED!
Name of created key: 00 0b 4f 63 f6 78 f9 44 70 1f 69 c0 d3 10 eb 07 
a1 d7 7e 3c cf b5 20 e6 4e 43 20 a4 e8 75 e2 68 
30 f4 
	passing case:  	PASSED!
Name of loading key: 00 0b 02 93 c8 ac 40 6a be 1b ea cc ef c5 1f c5 
07 c2 b5 18 97 64 5a d7 fa 16 15 cb 6f de d7 8c 
6f e9 

Loaded key handle:  8000000d
	passing case:  	PASSED!
Encrypted data: d3 4a 41 af 2d 04 a7 91 e8 22 68 51 88 5f 1b 63 
d6 26 56 f7 ef e5 69 99 85 9c 67 5d 15 54 f5 d4 
a7 be 01 bd eb a5 87 13 59 b5 a9 37 89 2d dd 43 
21 03 82 94 80 67 a5 7f 02 30 f3 14 62 ee 24 96 
39 2d 51 e8 0e 1b 05 99 cd 8c 2f 56 e0 27 fa 8c 
7c 67 cf 6a 8f 9a 9b 66 ec 7f 53 f5 b0 e2 fa 98 
ba dc a2 28 a7 39 2f 65 43 bd bd 4b fb a0 1e 4a 
72 26 42 a5 7f da 4a 92 21 57 8c 1b 25 d4 58 e3 
7f f8 fb 8f 04 62 e8 0b 38 c5 42 cf 33 d9 30 48 
0f 2c df 16 4b df b3 0d 50 f1 bd bb 9a fd f0 00 
cb 1d 0a 87 58 10 1f 1b 30 61 8e 30 d2 f5 17 05 
4f 14 df 3d 19 39 6c 82 da 8a d6 f6 a8 9c 22 8f 
f7 d3 6a dc f5 2d 07 35 e5 2e 1b 8b cc e9 5f ed 
3f fd 60 7d fe 54 f2 45 10 e9 cf e8 b9 1a 9b 5d 
ed dc 73 05 1a 32 fe dd 22 f6 de 84 1c d0 3a d5 
cc 86 60 16 a5 d1 f8 9c a0 da 1a ce 0e 72 d6 fa 

	passing case:  	PASSED!
Decypted data: 74 65 73 74 

HASH TESTS: HASH Comparison
	passing case:  	PASSED!
	passing case:  	PASSED!
Hash value before encrypt: 9f 86 d0 81 88 4c 7d 65 9a 2f ea a0 c5 5a d0 15 
a3 bf 4f 1b 2b 0b 82 2c d1 5d 6c 15 b0 f0 0a 08 

Hash value after decrypt: 9f 86 d0 81 88 4c 7d 65 9a 2f ea a0 c5 5a d0 15 
a3 bf 4f 1b 2b 0b 82 2c d1 5d 6c 15 b0 f0 0a 08 

But if running all tests automatically, test #34 - **FAIL**:
ASYMMETRIC ENCRYPT/DECRYPT & HASH TESTS:
	passing case:  	PASSED!

New key successfully created in owner hierarchy (RSA 2048).  Handle: 0x80000005
Name of created primary key: 00 0b 4f 63 f6 78 f9 44 70 1f 69 c0 d3 10 eb 07 
a1 d7 7e 3c cf b5 20 e6 4e 43 20 a4 e8 75 e2 68 
30 f4 
	passing case:  	FAILED!  System API Error: 0x80006

**Comment by yunyings** [comment]
Anyone could help with this or share comments?

**Comment by yunyings** [comment]
I just did same tests with latest kernel 4.7-rc5, same result - #34 PASS if runs alone, FAIL when all tests runs automatically.

**Comment by DamienMolina** [comment]
I have a problem with this test too(and not only this test), but I don't have the same error. I have
```
ASYMMETRIC ENCRYPT/DECRYPT TESTS:
Q - QUIT THIS TEST GROUP
D - PRINT DESCRIPTION ON ALL CASES IN THIS GROUP
0 - RUN ALL TEST CASES
Please select an action:0

RUN ALL TEST CASES:

ASYMMETRIC ENCRYPT/DECRYPT & HASH TESTS:
        passing case:   FAILED!  TPM Error: 0x9a2

```


Anyway, I believe the error 0x80006 is `TSS2_SYS_RC_INSUFFICIENT_BUFFER`  as described [here](https://github.com/01org/TPM2.0-TSS/blob/master/include/tss2/tss2_common.h#L192):
```
#define TSS2_SYS_RC_INSUFFICIENT_BUFFER         ((TSS2_RC)(TSS2_SYS_ERROR_LEVEL | \
                                                     TSS2_BASE_RC_INSUFFICIENT_BUFFER))
```

`#define TSS2_BASE_RC_INSUFFICIENT_BUFFER    6 /* A buffer isn't large enough */`

**Comment by yunyings** [comment]
Thanks Molina, and agreed. 

I also did some digging about the error code 0x80006:
asymmetricEncryptDecryptTest() -> Tss2_Sys_Create() ->... -> Unmarshal_Simple_TPM2B():
                length = CHANGE_ENDIAN_WORD( *(UINT16 *)*nextData );
 
                if( outTPM2B != 0 )
                {
                    if( length > outTPM2B->size )
                    {
                        *rval = TSS2_SYS_RC_INSUFFICIENT_BUFFER;
                    }  

It seems INSUFFICIENT_BUFFER is returned when resourcemanager calls Unmarshal_Simple_TPM2B() for test case #34 when runs all test cases automatically.
However, no such problem if runs #34 case alone. 

It seems to me like something goes wrong within the test tool when running all tests by select "0".
But need confirmation from TSS experts here.

**Comment by gwei3** [comment]
It should be a bug in tpmtest related to TPM2B_* initialization . I will dig it out and fix it.

**Comment by gwei3** [comment]
fixed in PR #234.
